raster/src/main/scala/geotrellis/raster/io/geotiff/GeoTiffMultibandTile.scala

```diff
@@ -321,8 +324,145 @@ abstract class GeoTiffMultibandTile(
+
+              case (_, UByteCellType) =>
+                (bytes: Array[Byte]) => new UByteRawGeoTiffSegment(bytes)
+              case (_, UByteConstantNoDataCellType) =>
+                (bytes: Array[Byte]) => new UByteConstantNoDataCellTypeGeoTiffSegment(bytes)
+              case (_, UByteUserDefinedNoDataCellType(nd)) =>
+                (bytes: Array[Byte]) => new UByteUserDefinedNoDataGeoTiffSegment(bytes, nd)
+
+              case (_, ShortCellType) =>
+                (bytes: Array[Byte]) => new Int16RawGeoTiffSegment(bytes)
+              case (_, ShortConstantNoDataCellType) =>
+                (bytes: Array[Byte]) => new Int16ConstantNoDataGeoTiffSegment(bytes)
+              case (_, ShortUserDefinedNoDataCellType(nd)) =>
+                (bytes: Array[Byte]) => new Int16UserDefinedNoDataGeoTiffSegment(bytes, nd)
+
+              case (_, UShortCellType) =>
+                (bytes: Array[Byte]) => new UInt16RawGeoTiffSegment(bytes)
+              case (_, UShortConstantNoDataCellType) =>
+                (bytes: Array[Byte]) => new UInt16ConstantNoDataGeoTiffSegment(bytes)
+              case (_, UShortUserDefinedNoDataCellType(nd)) =>
+                (bytes: Array[Byte]) => new UInt16UserDefinedNoDataGeoTiffSegment(bytes, nd)
+
+              case (_, IntCellType) =>
+                (bytes: Array[Byte]) => new Int32RawGeoTiffSegment(bytes)
+              case (_, IntConstantNoDataCellType) =>
+                (bytes: Array[Byte]) => new Int32ConstantNoDataGeoTiffSegment(bytes)
+              case (_, IntUserDefinedNoDataCellType(nd)) =>
+                (bytes: Array[Byte]) => new Int32UserDefinedNoDataGeoTiffSegment(bytes, nd)
+
+              case (_, FloatCellType) =>
+                (bytes: Array[Byte]) => new Float32RawGeoTiffSegment(bytes)
+              case (_, FloatConstantNoDataCellType) =>
+                (bytes: Array[Byte]) => new Float32ConstantNoDataGeoTiffSegment(bytes)
+              case (_, FloatUserDefinedNoDataCellType(nd)) =>
+                (bytes: Array[Byte]) => new Float32UserDefinedNoDataGeoTiffSegment(bytes, nd)
+
+              case (_, DoubleCellType) =>
+                (bytes: Array[Byte]) => new Float64RawGeoTiffSegment(bytes)
+              case (_, DoubleConstantNoDataCellType) =>
+                (bytes: Array[Byte]) => new Float64ConstantNoDataGeoTiffSegment(bytes)
+              case (_, DoubleUserDefinedNoDataCellType(nd)) =>
+                (bytes: Array[Byte]) => new Float64UserDefinedNoDataGeoTiffSegment(bytes, nd)
+            }
+
+          val bytesPerSample = bandType.bytesPerSample
+          val setter = if (cellType.isFloatingPoint) tileSetDouble else tileSetInt
+
+          getSegments(0 until segmentCount).foreach { case (segmentIndex, geoTiffSegment) =>
+            val segmentTransform = segmentLayout.getSegmentTransform(segmentIndex, bandCount)
+            val bytes = GeoTiffSegment.deinterleave(geoTiffSegment.bytes, bandCount, bytesPerSample, 0 until bandCount)
```

Comment:
I don't think it's necessary to de-interleave here, and causes more array iterations than necessary.

I think a good approach would be to iterate over the segments as-is, and write to the bands that each value should be. i.e. there's no reason why the band count has to be the outer most loop of the iteration.

Is there a reason why we couldn't call `getSegments` and foreach over the segments as the BandInterleave case, and then write the values to the appropriate place in the result bands?

Reply:
Would not just `crop(this.gridBounds)` be a desired optimisation?

Reply:
Yeah, simply doing `crop(this.gridBounds)` would work and is implemented in the desired way.